File: Sources/Kernel/MM/VMM.swift
/*
 * MM/VMM.swift
 * Virtual Memory Management
 */

import CSupport

// Global state for VMM
nonisolated(unsafe) private var pml4: UnsafeMutablePointer<UInt64>!

public struct VMM {
    public static func setup() {
        kprint("VMM setup\n")
        let cr3 = asm_get_cr3()
        let physPML4 = cr3 & ~0xFFF
        pml4 = UnsafeMutablePointer<UInt64>(bitPattern: UInt(physPML4))!
    }

    // Flags: 1=Present, 2=RW, 4=User
    public static func map(virt: UInt64, phys: PhysAddr, flags: UInt64) {
        let pml4Index = (virt >> 39) & 0x1FF
        let pdptIndex = (virt >> 30) & 0x1FF
        let pdIndex = (virt >> 21) & 0x1FF
        let ptIndex = (virt >> 12) & 0x1FF

        // PML4 -> PDPT
        guard let pdptPhys = getOrAllocTable(entry: &pml4[Int(pml4Index)]) else { return }
        let pdpt = UnsafeMutablePointer<UInt64>(bitPattern: UInt(pdptPhys))!

        // PDPT -> PD
        guard let pdPhys = getOrAllocTable(entry: &pdpt[Int(pdptIndex)]) else { return }
        let pd = UnsafeMutablePointer<UInt64>(bitPattern: UInt(pdPhys))!

        // PD -> PT
        guard let ptPhys = getOrAllocTable(entry: &pd[Int(pdIndex)], isL2: true) else { return }
        let pt = UnsafeMutablePointer<UInt64>(bitPattern: UInt(ptPhys))!

        // PT Entry
        pt[Int(ptIndex)] = phys.value | flags | 1  // Always Present
        asm_invlpg(UnsafeMutableRawPointer(bitPattern: UInt(virt)))
    }

    private static func getOrAllocTable(entry: UnsafeMutablePointer<UInt64>, isL2: Bool = false)
        -> UInt64?
    {
        var val = entry.pointee
        if (val & 1) == 0 {
            // Not present, allocate new table
            guard let frame = PMM.allocateFrame() else { return nil }
            // Default flags: User | RW | Present
            val = frame.value | 7
            entry.pointee = val
            return frame.value
        }

        // Check for Huge Page (Bit 7, value 0x80)
        if (val & 0x80) != 0 {
            if isL2 {
                return split2MBPage(entry: entry)
            }
            return nil
        }

        return val & ~0xFFF
    }

    private static func split2MBPage(entry: UnsafeMutablePointer<UInt64>) -> UInt64? {
        let hugePagePhys = entry.pointee & ~UInt64(0x1FFFFF)
        let hugePageFlags = entry.pointee & 0x1FF
        let ptFlags = hugePageFlags & ~UInt64(0x80)

        guard let frame = PMM.allocateFrame() else { return nil }
        let pt = UnsafeMutablePointer<UInt64>(bitPattern: UInt(frame.value))!

        for i in 0..<512 {
            pt[i] = (hugePagePhys + UInt64(i * 4096)) | ptFlags
        }

        // Update the PD entry to point to this new PT
        entry.pointee = frame.value | 7
        return frame.value
    }
}

File: zig-kernel/src/vmm.zig
const pmm = @import("pmm.zig");

pub var pml4: [*]u64 = undefined;

pub fn setup() void {
    var cr3: u64 = undefined;
    asm volatile ("mov %%cr3, %[cr3]"
        : [cr3] "=r" (cr3),
    );
    const phys_pml4: usize = @intCast(cr3 & ~@as(u64, 0xFFF));
    pml4 = @ptrFromInt(phys_pml4);
}

pub fn map(virt: u64, phys: usize, flags: u64) void {
    const pml4Index: usize = @intCast((virt >> 39) & 0x1FF);
    const pdptIndex: usize = @intCast((virt >> 30) & 0x1FF);
    const pdIndex: usize = @intCast((virt >> 21) & 0x1FF);
    const ptIndex: usize = @intCast((virt >> 12) & 0x1FF);

    // PML4 -> PDPT
    if (getOrAllocTable(&pml4[pml4Index], false) == null) return;
    const pdpt_phys: usize = @intCast(getOrAllocTable(&pml4[pml4Index], false) orelse return);
    const pdpt: [*]u64 = @ptrFromInt(pdpt_phys);

    // PDPT -> PD
    if (getOrAllocTable(&pdpt[pdptIndex], false) == null) return;
    const pd_phys: usize = @intCast(getOrAllocTable(&pdpt[pdptIndex], false) orelse return);
    const pd: [*]u64 = @ptrFromInt(pd_phys);

    // PD -> PT
    if (getOrAllocTable(&pd[pdIndex], true) == null) return;
    const pt_phys: usize = @intCast(getOrAllocTable(&pd[pdIndex], true) orelse return);
    const pt: [*]u64 = @ptrFromInt(pt_phys);

    // PT Entry
    pt[ptIndex] = @as(u64, phys) | flags | 1; // Present

    const addr_ptr: *const u8 = @ptrFromInt(@as(usize, @intCast(virt)));
    asm volatile ("invlpg (%[addr])"
        :
        : [addr] "r" (addr_ptr),
    );
}

fn getOrAllocTable(entry: *u64, isL2: bool) ?usize {
    const val: u64 = entry.*;
    if ((val & 1) == 0) {
        if (pmm.allocateFrame()) |frame| {
            const v = @as(u64, frame) | 7; // Present|RW|User
            entry.* = v;
            return frame;
        } else {
            return null;
        }
    }

    // Huge page check (PS bit at bit 7 => 0x80 in entry flags for PD)
    if ((val & 0x80) != 0) {
        if (isL2) {
            return split2MBPage(entry);
        }
        return null;
    }

    return @intCast(val & ~(@as(u64, 0xFFF)));
}

fn split2MBPage(entry: *u64) ?usize {
    const hugePagePhys: u64 = entry.* & ~(@as(u64, 0x1FFFFF));
    const hugePageFlags: u64 = entry.* & @as(u64, 0x1FF);
    const ptFlags: u64 = hugePageFlags & ~(@as(u64, 0x80));

    const frame = pmm.allocateFrame() orelse return null;
    const pt: [*]u64 = @ptrFromInt(frame);

    var i: usize = 0;
    while (i < 512) : (i += 1) {
        pt[i] = hugePagePhys + @as(u64, i * 4096) | ptFlags;
    }

    entry.* = @as(u64, frame) | 7;
    return frame;
}
